google/cloud/pubsublite/internal/topic_partition_count_reader_impl_test.cc

```diff
@@ -0,0 +1,81 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/pubsublite/internal/topic_partition_count_reader_impl.h"
+#include "google/cloud/pubsublite/mocks/mock_admin_connection.h"
+#include "google/cloud/future.h"
+#include "google/cloud/status_or.h"
+#include "google/cloud/testing_util/is_proto_equal.h"
+#include <gmock/gmock.h>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace pubsublite_internal {
+
+using ::google::cloud::testing_util::IsProtoEqual;
+
+using google::cloud::pubsublite::Topic;
+using google::cloud::pubsublite::v1::GetTopicPartitionsRequest;
+using google::cloud::pubsublite::v1::TopicPartitions;
+using google::cloud::pubsublite_mocks::MockAdminServiceConnection;
+
+TEST(TestAsyncRead, Valid) {
+  std::uint32_t const num_partitions = 50;
+  auto connection = std::make_shared<MockAdminServiceConnection>();
+  TopicPartitionCountReaderImpl reader{connection};
+  Topic topic{"project", "location", "name"};
+  GetTopicPartitionsRequest req;
+  *req.mutable_name() = topic.FullName();
+  auto p = std::make_shared<promise<void>>();
```

Comment:
no need to use shared_ptr in tests, just put it on the stack and capture everything by reference
[&](GetTopicPartitionsRequest...